### PR TITLE
[CALCITE-6262] CURRENT_TIMESTAMP(P) ignores DataTypeSystem#getMaxPrecision

### PR DESCRIPTION
**Pull request by jduo**
Change datetime functions taking in a precision to validate against the max precision from the type system, rather than the SqlTypeName.

### CODE REVIEW

**Comment by jduo** [comment]
@tanclary

File: core/src/test/java/org/apache/calcite/sql/type/RelDataTypeSystemTest.java
Comment (tanclary) [inline core/src/test/java/org/apache/calcite/sql/type/RelDataTypeSystemTest.java:132]: Where is the +3 coming from?

File: core/src/test/java/org/apache/calcite/sql/type/RelDataTypeSystemTest.java
Comment (tanclary) [inline core/src/test/java/org/apache/calcite/sql/type/RelDataTypeSystemTest.java:237]: would be helpful to leave a comment with the jira case. there are lots of examples with `Test case for ...` you can use

File: core/src/test/java/org/apache/calcite/sql/type/RelDataTypeSystemTest.java
Comment (tanclary) [inline core/src/test/java/org/apache/calcite/sql/type/RelDataTypeSystemTest.java:131]: why not just have a `MAX_TIMESTAMP_PRECISION`? Am I wrong in saying that everytime we look for the max timestamp precision we're just retrieving the max datetime and then adding 3? In that case we should just define it as a constant and then this becomes unnecessary.

**Review by tanclary — CHANGES_REQUESTED** [review]

File: core/src/test/java/org/apache/calcite/sql/type/RelDataTypeSystemTest.java
Comment (jduo) [reply core/src/test/java/org/apache/calcite/sql/type/RelDataTypeSystemTest.java:132]: It's just a made-up number that differs from the default max.

File: core/src/test/java/org/apache/calcite/sql/type/RelDataTypeSystemTest.java
Comment (jduo) [reply core/src/test/java/org/apache/calcite/sql/type/RelDataTypeSystemTest.java:133]: I named this constant CUSTOM_MAX_TIMESTAMP_PRECISION just to be explicit that it's testing a custom max.

File: core/src/test/java/org/apache/calcite/sql/type/RelDataTypeSystemTest.java
Comment (jduo) [reply core/src/test/java/org/apache/calcite/sql/type/RelDataTypeSystemTest.java:244]: Done

**Comment by jduo** [comment]
@tanclary , the linter fixes are done. Please let me know if there are any other concerns.

**Review by tanclary — APPROVED** [review]

File: core/src/test/java/org/apache/calcite/sql/type/RelDataTypeSystemTest.java
Comment (JiajunBernoulli) [inline core/src/test/java/org/apache/calcite/sql/type/RelDataTypeSystemTest.java:52]: I don't understand this, I'm curious why it is + `3`?

Comment (tanclary) [reply core/src/test/java/org/apache/calcite/sql/type/RelDataTypeSystemTest.java:52]: I think he just chose an arbitrary number

Comment (jduo) [reply core/src/test/java/org/apache/calcite/sql/type/RelDataTypeSystemTest.java:52]: Yes, just the default offset by an arbitrary number (to ensure it is distinct from the default).

**Comment by jduo** [comment]
Hi, just wondering if there's anything needed on my side to get this one merged.
Thanks for the reviews.

**Comment by mihaibudiu** [comment]
If there are no comments before the end of the day I plan to merge this PR, since it seems to have an approval.

**Pull request merged**
